File: split/getnode.pas
function getnode(s: integer): halfword;
label
  40, 10, 20;
var
  p: halfword;
  q: halfword;
  r: integer;
  t: integer;
begin
  20:
    p := rover;
  repeat
    {127:}
q := p + mem[p].hh.lh;
    while (mem[q].hh.rh = 65535) do
    begin
      t := mem[q + 1].hh.rh;
      if q = rover then
        rover := t;
      mem[t + 1].hh.lh := mem[q + 1].hh.lh;
      mem[mem[q + 1].hh.lh + 1].hh.rh := t;
      q := q + mem[q].hh.lh;
    end;
    r := q - s;
    if r > p + 1 then{128:}

    begin
      mem[p].hh.lh := r - p;
      rover := p;
      goto 40;
    end{:128}
;
    if r = p then
      if mem[p + 1].hh.rh <> p then{129:}

      begin
        rover := mem[p + 1].hh.rh;
        t := mem[p + 1].hh.lh;
        mem[rover + 1].hh.lh := t;
        mem[t + 1].hh.rh := rover;
        goto 40;
      end{:129}
;
    mem[p].hh.lh := q - p{:127}
;
    p := mem[p + 1].hh.rh;
  until p = rover;
  if s = 1073741824 then
  begin
    getnode := 65535;
    goto 10;
  end;
  if lomemmax + 2 < himemmin then
    if lomemmax + 2 <= 65535 then{126:}

    begin
      if himemmin - lomemmax >= 1998 then
        t := lomemmax + 1000
      else
        t := lomemmax + 1 + (himemmin - lomemmax) div 2;
      p := mem[rover + 1].hh.lh;
      q := lomemmax;
      mem[p + 1].hh.rh := q;
      mem[rover + 1].hh.lh := q;
      if t > 65535 then
        t := 65535;
      mem[q + 1].hh.rh := rover;
      mem[q + 1].hh.lh := p;
      mem[q].hh.rh := 65535;
      mem[q].hh.lh := t - lomemmax;
      lomemmax := t;
      mem[lomemmax].hh.rh := 0;
      mem[lomemmax].hh.lh := 0;
      rover := q;
      goto 20;
    end{:126}
;
  overflow(300, memmax + 1 - memmin);
  40:
    mem[r].hh.rh := 0;{varused:=varused+s;}

  getnode := r;
  10: ;
end;

File: cpp/src/getnode.cpp
#include "getnode.h"
#include "erreur.h"

halfword getnode(int s)
{
	bool label20;
	halfword r;
	do
	{
		label20 = false;
		halfword p = rover;
		do
		{ 
			halfword q = p+info(p);
			while (link(q) == empty_flag)
			{
				auto t = link(q+1);
				if (q == rover)
					rover = t;
				info(t+1) = info(q+1);
				link(info(q+1)+1) = t;
				q += info(q);
			}
			r = q-s;
			if (r > p+1)
			{
				info(p) = r-p;
				rover = p;
				link(r) = 0;
				return r;
			}
			if (r == p)
			if (link(p+1) != p)
			{
				rover = link(p+1);
				int t = info(p+1);
				info(rover+1) = t;
				link(t+1) = rover;
				link(r) = 0;
				return r;
			}
			info(p) = q-p;
			p = link(p+1);
		} while (p != rover);
		if (s == 0x40'00'00'00)
			return empty_flag;
		if (lomemmax+2 < himemmin && lomemmax+2 <= empty_flag)
		{
			int t;
			if (himemmin-lomemmax >= 1998)
				t = lomemmax+1000;
			else
				t = lomemmax+1+(himemmin-lomemmax)/2;
			p = info(rover+1);
			halfword q = lomemmax;
			link(p+1) = q;
			info(rover+1) = q;
			if (t > empty_flag)
				t = empty_flag;
			link(q+1) = rover;
			info(q+1) = p;
			link(q) = empty_flag;
			info(q) = t-lomemmax;
			lomemmax = t;
			link(lomemmax) = 0;
			info(lomemmax) = 0;
			rover = q;
			label20 = true;
		}
	} while (label20);
	overflow("main memory size", memmax+1-memmin);
	link(r) = 0;
	return r;
}
